doc/source/ray-observability/monitoring-debugging/troubleshoot-failures.rst

```diff
@@ -1,44 +1,252 @@
 
-- **Stressful workloads:** It is currently possible for a process to get
-  marked as dead without actually having died. For example, workloads that
-  create many many tasks in a short amount of time can sometimes interfere
-  with the heartbeat mechanism that we use to check that processes are still alive.
-  On the head node in the cluster, you can check the files
-  ``/tmp/ray/session_*/logs/monitor*``. They will indicate which processes Ray has marked
-  as dead (due to a lack of heartbeats).
+Ray APIs are future APIs (indeed, it is :ref:`possible to convert Ray object references to standard Python future APIs <async-ref-to-futures>`), 
+and the error handling model is the same. When any remote tasks or actors fail, the returned object ref will contain an exception. 
+When you call ``get`` API to the object ref, it raises an exception.
 
-- **Starting many actors:** Workloads that start a large number of actors all at
-  once may exhibit problems when the processes (or libraries that they use)
-  contend for resources. Similarly, a script that starts many actors over the
-  lifetime of the application will eventually cause the system to run out of
-  file descriptors. This is addressable, but currently we do not garbage collect
-  actor processes until the script finishes.
+.. code-block:: python
 
-- **Running out of file descriptors:** As a workaround, you may be able to
-  increase the maximum number of file descriptors with a command like
-  ``ulimit -n 65536``. If that fails, double check that the hard limit is
-  sufficiently large by running ``ulimit -Hn``. If it is too small, you can
-  increase the hard limit as follows (these instructions work on EC2).
+  import ray
+  @ray.remote
+  def f():
+      raise ValueError
+  
+  # Raises a ValueError.
+  ray.get(f.remote())
 
-    * Increase the hard ulimit for open file descriptors system-wide by running
-      the following.
+In Ray, there are 3 types of failures. See exception APIs for more details. 
 
-      .. code-block:: bash
+- **Application failures**: This means the remote task/actor fails by the user code. In this case, ``get`` API will raise the ``RayTaskError`` which includes the exception raised from the remote process.
+- **Intentional system failures**: This means Ray is failed, but the failure is intended. For example, when you call cancellation APIs like ``ray.cancel`` (for task) or ``ray.kill`` (for actors), the system fails remote tasks and actors, but it is intentional.
+- **Unintended system failures**: This means the remote tasks and actors failed due to unexpected system failures such as processes crashing (for example, by out-of-memory error) or nodes failing. There are 4 common types.
+  1. `Linux Out of Memory killer <https://www.kernel.org/doc/gorman/html/understand/understand016.html>`_ or :ref:`Ray OOM killer <ray-oom-monitor>` kills processes with high memory usages to avoid out-of-memory.
```

Comment:
don't think the bullet points are showing up: 
<img width="569" alt="image" src="https://user-images.githubusercontent.com/11676094/232117743-a32bc21b-b03b-4725-ac44-36559e9d25eb.png">

Reply:
and there's only 3?

Reply:
Fixed both problems!